code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingCacheUtilities.java

```diff
@@ -0,0 +1,304 @@
+     *
+     * @param cachedFile {@code File} to which the content has to be written
+     * @param input      The {@code InputStream} to be written to the cache
+     * @param append     true, if you want to append the input stream to the existing file content
+     * @return {@code boolean} containing true if the {@code InputStream} has been successfully written into the file, false otherwise
+     */
+    private boolean readInputStreamIntoFile(final File cachedFile, final InputStream input, final boolean append) {
+        if (cachedFile == null || input == null) {
+            return false;
+        }
+
+        FileOutputStream output = null;
+
+        try {
+            output = new FileOutputStream(cachedFile, append);
+            final byte[] data = new byte[STREAM_WRITE_BUFFER_SIZE];
+            int count;
+
+            while ((count = input.read(data)) != -1) {
+                output.write(data, 0, count);
+            }
+        } catch (final IOException e) {
+            Log.error(LOG_TAG, "%s - IOException while attempting to write remote file (%s)", SELF_TAG, e);
+            return false;
+        } finally {
+            try {
+                if (output != null) {
+                    output.close();
+                }
+            } catch (final IOException e) {
+                Log.error(LOG_TAG, "%s - Unable to close the OutputStream (%s) ", SELF_TAG, e);
+            }
+        }
+
+        return true;
+    }
+
+    // ========================================================================================================
+    // Image asset caching
+    // ========================================================================================================
+
+    /**
+     * Caches the assets provided in the {@link java.util.List}.
+     *
+     * @param assetsCollection a {@link List<String>} containing asset url's to be cached.
+     */
+    void cacheImageAssets(final List<String> assetsCollection) {
+        if (cacheManager == null) {
+            Log.trace(LOG_TAG, "%s - Failed to cache asset, the cache manager is not available.", SELF_TAG);
+            return;
+        }
+
+        final ArrayList<String> assetsToRetain = new ArrayList<>();
+
+        if (assetsCollection != null && !assetsCollection.isEmpty()) {
+            for (final String imageAsset : assetsCollection) {
+                if (assetIsDownloadable(imageAsset) && !assetsToRetain.contains(imageAsset)) {
+                    assetsToRetain.add(imageAsset);
+                }
+            }
+        }
+
+        // clean any cached images which are no longer needed then use the RemoteDownloader to download
+        // any new assets that have not been previously cached.
+        cacheManager.deleteFilesNotInList(assetsToRetain, IMAGES_CACHE_SUBDIRECTORY);
+
+        for (final String asset : assetsToRetain) {
+            try {
+                final RemoteDownloader remoteDownloader = getRemoteDownloader(asset, IMAGES_CACHE_SUBDIRECTORY);
```

Comment:
Is there a possibility that asset is already cached. Since in line 229 we are removing all the files that are not part of assetsToRetain.

Reply:
there is a possibility that this is true and in that case the remote downloader will use the If-Modified header when making the download request.